keps/sig-network/2943-networkpolicy-status/README.md

```diff
@@ -0,0 +1,641 @@
+Samantha, the cluster admin of a Kubernetes cluster is watching the complaints about Network 
+Policies not working in the cluster raise after she upgraded the cluster CNI to the latest 
+version.
+
+Users decided to create Network Policies with some fancy new features that were announced 
+in the release notes of a random CNI, not knowing that the CNI used in Samantha's cluster 
+is different and still cannot implement the new fields.
+
+Samantha needs to know exactly what are the complaints and why the CNI cannot parse the 
+new rules, so she can explain to the users and also plan their monthly cluster upgrade to 
+add the required features.
+
+
+### Notes/Constraints/Caveats (Optional)
+The implementation of a status field depends on the CNI to provide this feedback. It's 
+not up to Kubernetes to know what happened, but to the controller of that object.
+
+If decided to use a "generic" approach of "controller / status / description" we cannot 
+guarantee that the CNI providers will follow the same standard/defaulting of the messages, 
+and this might be frustrating for users.
+
+### Risks and Mitigations
+Wrong RBAC in .status subresource field may lead to someone adding wrong status to users, and this
+may get mitigated by CNIs when using the right RBAC that allows only their "controllers" to change 
+this subresource.
+
+## Design Details
+API changes to NetworkPolicy:
+* Add the `status` inside `NetworkPolicy` struct as following:
+```go
+// NetworkPolicy describes what network traffic is allowed for a set of Pods
+type NetworkPolicy struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired behavior for this NetworkPolicy.
+	// +optional
+	Spec NetworkPolicySpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+  // Status is the current state of the NetworkPolicy.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Status NetworkPolicyStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+```
+
+* Add the `NetworkPolicyStatus` struct as following:
+```go
+// NetworkPolicyStatus describe the current state of the NetworkPolicy.
+type NetworkPolicyStatus struct {
+	PolicyStatus []CNIPolicyStatus
+}
+```
+
+* Add CNIPolicyStatus type struct as the following:
+
+```go
+type CNIPolicyStatus struct {
+  // CNI defines the CNI name
+	CNI string
+  
+  // Status defines the status string (parsed, error, etc)
+  Status string
+
+  // Description defines the description of the Status (why it's on that status)
+  Description string
+}
+```
+
+```
+<<[UNRESOLVED optional short context or usernames ]>>
+We need to define if .status.PolicyStatus[].Status will have its own type with
```

Comment:
Will the properly added network policy have a "GOOD" status like the *parsed* by default.. Or the idea is to reserve this for errors only.

Reply:
Could be good to have the possible status list as an enumeration with fixed values.